source/common/runtime/runtime_impl.cc

```diff
@@ -555,9 +555,44 @@ void LoaderImpl::loadNewSnapshot() {
   tls_->set([ptr = std::move(ptr)](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
     return ptr;
   });
+  // Invalidate all locally tracked snapshots. These can not be deleted as the owning threads may
+  // be accessing them. Instead mark them invalid and clean up during the next in-thread call to
+  // snapshot()
+  {
+    absl::ReaderMutexLock lock(&snapshot_mutex_);
```

Comment:
I think this needs to be the write lock?

Also, the way this works as written is that if there are a lot of threads that are coming and going, we will potentially leak the map. This might be OK, but can you comment this for now?

As an aside, it would be better if snapshot() just returned the shared_ptr and then I think we could avoid all of this valid stuff and even the map of thread IDs. Thinking about it more, how about just creating a snapshot() variant that returns the shared pointer and using that in the static runtime lookup functions? If we have that, I believe all of this can be simplified a huge amount by just grabbing that shared pointer under read lock, and then writing the shared pointer under write lock in this function?

Reply:
yeah, should be a write lock.
Happy to comment regarding lifetime. I had an iteration which on invalidation would simply do a 5s delay delete, but was uncomfortable about lifetime not being fixed.

I started in on snapshot as shared pointer, and the thing I don't like about that is that the TLS snapshot isn't a shared pointer, so runtime feature lookup would always return the shared pointer snapshot not the TLS snapshot, which could arguably have different values during update.  I don't think there's anything unsafe about it, but it does mean we're doing more shared pointer access and locks, and runtime thrashing makes me twitchy.   Any thoughts on how to avoid that ugliness or you think it's the way to go?

Reply:
> the TLS snapshot isn't a shared pointer

The TLS snapshot is actually a shared pointer so I don't think anything needs to change in that path (just use the shared_ptr version of the TLS lookup function if it's registered). Let's sync up on slack?